compiler/rustc_codegen_cranelift/src/common.rs

```diff
@@ -336,7 +336,7 @@ impl<'tcx> HasTargetSpec for FunctionCx<'_, '_, 'tcx> {
 impl<'tcx> FunctionCx<'_, '_, 'tcx> {
     pub(crate) fn monomorphize<T>(&self, value: T) -> T
     where
-        T: TypeFoldable<'tcx> + Copy,
+        T: TypeFoldable<TyCtxt<'tcx>> + Copy,
```

Comment:
Could there be a trait alias for TypeFoldable filling in TyCtxt in rustc_middle? That would significantly reduce the diff.

Reply:
Wasn't aware of trait aliases!  Good idea, will look into it.

Reply:
Okay so, whilst a trait alias can be used in bounds such as these, it can't be used to implement the trait: so we'll still need `impl<'tcx> TypeFoldable<TyCtxt<'tcx>> for ...` followed by `... where T: TypeFoldableAlias<'tcx>`—and hence we'd still end up modifying each of these bounds to refer to the alias instead, unless the alias is named `TypeFoldable` (which I think would be very confusing unless the trait itself is renamed something else?).  So the diff will likely remain similar, but the boilerplate would be slightly reduced.

Reply:
I think it's fine to have a generic TypeFoldable that is almost never used and one that is used everywhere else. We do a similar thing for `Ty` after all. It's just that type aliases are stable.

Reply:
I created [another branch](https://github.com/eggyal/rust/tree/move_fold_visit_traits_to_type_lib_with_type_alias) with an alternative implementation using type aliases, but as suspected the impact on the overall diff is not so significant: +1494-1309 across 99 files versus this PR's +1709-1434 across 112 files.

I'm happy to use/force-push that if it's preferred, but IMHO defining e.g. `TypeFoldable` to mean different things depending on whether the alias or the underlying trait are brought into scope is pretty confusing (and all the existing reexports certainly don't help to keep track of what each name means in a given context).  In particular for each trait, that branch now has at very least the following:

```rust
rustc_type_ir::fold::TypeFoldable // (1): actual trait definition

rustc_middle::ty::fold::TypeFoldable // (2): re-export of (1)
rustc_middle::ty::ir::TypeFoldable // (3): re-export of (2)

rustc_middle::ty::fold::ctxt::TypeFoldable // (4): type alias definition
rustc_middle::ty::TypeFoldable // (5): re-export of (4)
```

And the codebase variously uses each of the above in different and inconsistent ways.  Certainly this could (and probably should) be rationalised, but fundamentally I think the naming conflict is unhelpful.

Incidentally, you may note that branch also removes some previously required imports from scope: this is a direct result of using type aliases, which erroneously bring supertraits into scope (reported #107747).  Once that bug is resolved, many of those imports will need to be re-added.

Reply:
This kind of "shadowing" is what we do for types (we don't use `ir::TyKind<TyCtxt<'tcx>>` everywhere, but `rustc_middle::...::TyKind<'tcx>`), why would we do it differently for traits? To avoid any trait alias issues, we should be having two entirely separate traits of the same name. The rustc_middle trait (`TypeVisitable<'tcx>`) can be implemented for all types implementing `ir::TypeVisitable<TyCtxt<'tcx>>`, thus picking up any generic impls from earlier crates.

Reply:
Understood.  I think I will close this PR and open another using that approach.

Reply:
Thanks! ❤️

Reply:
On reflection, I'm not sure it's possible to blanket implement each (existing) `middle` trait (parameterised by `'tcx`) for all types implementing the respective `ir` trait (parameterised by `TyCtxt<'tcx>`), because generic methods in the `middle` traits (e.g. `TypeVisitable::visit_with`) are parameterised by a type that must implement some other `middle` trait (e.g. `V: TypeVisitor<'tcx>`) and, in delegating to the `ir` trait, the blanket implementation must pass a value of that generic type to its equivalent method (`ir::TypeVisitable::visit_with`), which requires that value also to implement the respective `ir` trait (`ir::TypeVisitor<TyCtxt<'tcx>>`).  However this cannot be guaranteed unless the `middle` traits are subtraits of the `ir` traits, thus always requiring that the `ir` traits be implemented and negating much benefit of retaining the `middle` traits; any remaining ergonomic benefit is then negated by ambiguities that such supertraits introduce when resolving associated items.

Reply:
Hmm... I see. I'm assuming trait aliases won't allow `impls TraitAlias for Type`, thus leading to the branch you linked that still references `TypeVisitable<TyCtxt<'tcx>>` almost as often as before.

I'm convinced to reopen this PR, even if it's a bit annoying to have to do that all over the place.

Reply:
Thanks doing the experiment. Please add this information to the PR so that it's clear why this design was chosen

Reply:
Indeed, you cannot (currently?) `impl TraitAlias for ...`.

Will open a new PR (am looking into some other changes anyway) and provide the detail as suggested.  Thanks!